File: engine/src/main/java/com/terralite/engine/physics/Collider.java
package com.terralite.engine.physics;

public record Collider(double halfWidth, double halfHeight, double halfDepth) {
    public Collider {
        if (halfWidth < 0.0) {
            throw new IllegalArgumentException("Collider half width cannot be negative");
        }
        if (halfHeight < 0.0) {
            throw new IllegalArgumentException("Collider half height cannot be negative");
        }
        if (halfDepth < 0.0) {
            throw new IllegalArgumentException("Collider half depth cannot be negative");
        }
    }

    public Aabb boundsAt(Transform transform) {
        return new Aabb(
            transform.x() - halfWidth,
            transform.y() - halfHeight,
            transform.z() - halfDepth,
            transform.x() + halfWidth,
            transform.y() + halfHeight,
            transform.z() + halfDepth
        );
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/physics/Collider.kt
package com.terralite.engine.physics

@JvmRecord
data class Collider(val halfWidth: Double, val halfHeight: Double, val halfDepth: Double) {
    init {
        require(halfWidth >= 0.0) { "Collider half width cannot be negative" }
        require(halfHeight >= 0.0) { "Collider half height cannot be negative" }
        require(halfDepth >= 0.0) { "Collider half depth cannot be negative" }
    }

    fun boundsAt(transform: Transform): Aabb = Aabb(
        transform.x - halfWidth,
        transform.y - halfHeight,
        transform.z - halfDepth,
        transform.x + halfWidth,
        transform.y + halfHeight,
        transform.z + halfDepth
    )
}
